src/ordered_robin_dict.jl

```diff
@@ -0,0 +1,476 @@
+
+```jldoctest
+julia> OrderedRobinDict("A"=>1, "B"=>2)
+OrderedRobinDict{String,Int64} with 2 entries:
+  "A" => 1
+  "B" => 2
+```
+"""
+mutable struct OrderedRobinDict{K,V} <: AbstractDict{K,V}
+    dict::RobinDict{K, Int32} 
+    keys::Vector{K}
+    vals::Vector{V}
+    count::Int32
+
+    function OrderedRobinDict{K, V}() where {K, V}
+        new{K, V}(RobinDict{K, Int32}(), Vector{K}(), Vector{V}(), 0)
+    end
+
+    function OrderedRobinDict{K, V}(d::OrderedRobinDict{K, V}) where {K, V}
+        new{K, V}(copy(d.dict), copy(d.keys), copy(d.vals), d.count)
+    end
+
+    function OrderedRobinDict{K,V}(kv) where {K, V}
+        h = OrderedRobinDict{K,V}()
+        for (k,v) in kv
+            h[k] = v
+        end
+        return h
+    end
+    OrderedRobinDict{K,V}(p::Pair) where {K,V} = setindex!(OrderedRobinDict{K,V}(), p.second, p.first)
+    function OrderedRobinDict{K,V}(ps::Pair...) where V where K
+        h = OrderedRobinDict{K,V}()
+        sizehint!(h, length(ps))
+        for p in ps
+            h[p.first] = p.second
+        end
+        return h
+    end
+end
+
+OrderedRobinDict() = OrderedRobinDict{Any,Any}()
+OrderedRobinDict(kv::Tuple{}) = OrderedRobinDict()
+copy(d::OrderedRobinDict) = OrderedRobinDict(d)
+empty(d::OrderedRobinDict, ::Type{K}, ::Type{V}) where {K, V} = OrderedRobinDict{K, V}()
+
+OrderedRobinDict(ps::Pair{K,V}...) where {K,V} = OrderedRobinDict{K,V}(ps)
+OrderedRobinDict(ps::Pair...)                  = OrderedRobinDict(ps)
+
+OrderedRobinDict(d::AbstractDict{K, V}) where {K, V} = OrderedRobinDict{K, V}(d)
+
+function OrderedRobinDict(kv)
+    try
+        return dict_with_eltype((K, V) -> OrderedRobinDict{K, V}, kv, eltype(kv))
+    catch e
+        if !isiterable(typeof(kv)) || !all(x -> isa(x, Union{Tuple,Pair}), kv)
+            !all(x->isa(x,Union{Tuple,Pair}),kv)
+            throw(ArgumentError("OrderedRobinDict(kv): kv needs to be an iterator of tuples or pairs"))
+        else
+            rethrow(e)
+        end
+    end
+end
+
+empty(d::OrderedRobinDict{K,V}) where {K,V} = OrderedRobinDict{K,V}()
+
+length(d::OrderedRobinDict) = d.count
+isempty(d::OrderedRobinDict) = (length(d) == 0)
+
+"""
+    empty!(collection) -> collection
+
+Remove all elements from a `collection`.
+
+# Examples
+```jldoctest
+julia> A = OrderedRobinDict("a" => 1, "b" => 2)
+OrderedRobinDict{String,Int64} with 2 entries:
+  "a" => 1
+  "b" => 2
+
+julia> empty!(A);
+
+julia> A
+OrderedRobinDict{String,Int64} with 0 entries
+```
+"""
+function empty!(h::OrderedRobinDict{K,V}) where {K, V}
+    empty!(h.dict)
+    empty!(h.keys)
+    empty!(h.vals)
+    h.count = 0
+    return h
+end
+
+function _setindex!(h::OrderedRobinDict, v, key)
+    hk, hv = h.keys, h.vals
+    push!(hk, key)
+    push!(hv, v)
+    nk = length(hk)
+    @inbounds h.dict[key] = Int32(nk)
+    h.count += 1
+end
+
+function setindex!(h::OrderedRobinDict{K, V}, v0, key0) where {K,V}
+    key = convert(K,key0)
+
+    v = convert(V,  v0)
+
```

Comment:
is the extra whitespace here for a reason?
```suggestion
    v = convert(V,  v0)
```